backend/dal/Helpers/Extensions/ResearchExtensions.cs

```diff
@@ -0,0 +1,140 @@
+            }
+
+            if (filter.UpdatedOnEndDate.HasValue)
+            {
+                query = query.Where(l => (l.AppLastUpdateTimestamp.Date <= filter.UpdatedOnEndDate.Value.Date));
+            }
+
+            if (!string.IsNullOrWhiteSpace(filter.CreatedByIdir))
+            {
+                query = query.Where(r => EF.Functions.Like(r.AppCreateUserid, $"%{filter.CreatedByIdir}%"));
+            }
+
+            if (!string.IsNullOrWhiteSpace(filter.UpdatedByIdir))
+            {
+                query = query.Where(r => EF.Functions.Like(r.AppLastUpdateUserid, $"%{filter.UpdatedByIdir}%"));
+            }
+
+            if (filter.Sort?.Any() == true)
+            {
+                query = query.OrderByProperty(filter.Sort);
+            }
+            else
+            {
+                query = query.OrderBy(l => l.RfileNumber);
+            }
+
+            return query.Include(r => r.ResearchFileStatusTypeCodeNavigation)
+                .Include(r => r.PimsPropertyResearchFiles)
+                .ThenInclude(p => p.Property)
+                .ThenInclude(p => p.RegionCodeNavigation);
+        }
+
+        /// <summary>
+        /// Generate a query for the specified 'filter'.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="filter"></param>
+        /// <returns></returns>
+        public static IQueryable<Entities.PimsResearchFile> GenerateResearchQuery(this PimsContext context, Entity.Models.ResearchFilter filter)
+        {
+            filter.ThrowIfNull(nameof(filter));
+
+            var query = context.PimsResearchFiles.AsNoTracking();
+
+            query = query.GenerateCommonResearchQuery(filter);
+
+            return query;
+        }
+
+        /// <summary>
+        /// Get the next available id from the PIMS_RESEARCH_FILE_ID_SEQ
+        /// </summary>
+        /// <param name="context"></param>
+        public static Int64 GetNextResearchFileSequenceValue(this PimsContext context)
+        {
+            SqlParameter result = new SqlParameter("@result", System.Data.SqlDbType.BigInt)
+            {
+                Direction = System.Data.ParameterDirection.Output
+            };
+            context.Database.ExecuteSqlRaw("set @result = next value for dbo.PIMS_RESEARCH_FILE_ID_SEQ;", result);
+
+            return (Int64)result.Value;
+        }
+
+        /// <summary>
+        /// Generate a new R File in format R-XXX-YYY using the research file id. Add the research file id and rfileNo to the passed research file.
+        /// </summary>
+        /// <param name="context"></param>
+        public static PimsResearchFile GenerateRFileNo(this PimsContext context, PimsResearchFile researchFile)
+        {
+            Int64 researchFileId = GetNextResearchFileSequenceValue(context);
+            researchFile.ResearchFileId = researchFileId;
+            researchFile.RfileNumber = $"L-{researchFile.ResearchFileId.ToString().PadLeft(6, '0').Insert(3, "-")}";
```

Comment:
As per comments this should generate R-XXX-YY identifiers. However I see it being generated with "L-XXX-YY"  - perhaps a typo?

Reply:
this was refactored to be part of the research repository, I can remove it.